lib/mayaUsd/undo/UsdUndoManager.cpp

```diff
@@ -29,9 +29,16 @@ UsdUndoManager& UsdUndoManager::instance()
     return undoManager;
 }
 
-void UsdUndoManager::trackLayerStates(const SdfLayerHandle& layer)
+void UsdUndoManager::trackStatesOnNewStage(const SdfLayerHandle& layer)
 {
-    layer->SetStateDelegate(UsdUndoStateDelegate::New());
+    _layerStateDelegate = TfCreateRefPtr(new UsdUndoStateDelegate);
```

Comment:
Why do we save this as a data member?  Since the UsdUndoManager is a singleton, doesn't that mean that more than one layer can share the same UsdUndoStateDelegate?

Thinking out loud here, could it be that the UsdUndoManager should track a set of layers, and when it sees one that isn't in its set, whether than be through a layer creation notification or an edit target change notification, it will give the unknown layer a newly-created UsdUndoStateDelegate, then will add the layer to its set of known layers?

Reply:
> Why do we save this as a data member? 

The main reason I need to save _layerStateDelegate as a data member is for undo/redo purposes between layers itself.


> Since the UsdUndoManager is a singleton, doesn't that mean that more than one layer can share the same UsdUndoStateDelegate?

Exactly, more than one layer can share the same UsdUndoStateDelegate in a same stage.

> it will give the unknown layer a newly-created UsdUndoStateDelegate, then will add the layer to its set of known layers?

Let's have a chat around this. I would like to understand this approach better.